### ISSUE TITLE
Improve error models

### ISSUE THREAD

**Issue by epau** [awslabs/aws-sdk-swift#692]
### Describe the feature

Currently, it can be cumbersome for customers to extract info from the specific error that is being thrown. This is caused by a nesting of errors such as `SdkError.someError(ClientError.someError(TheActualError))`

Let's design a better error models so that the specific error thrown is easily accessible and reduce any unnecessary wrapping. 

### Use Case

N/A

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

**Comment by shepazon** [comment]
Yes please!

**Comment by shepazon** [comment]
There are two primary issues with this for me:

1. Exceptions thrown that contain nested objects describing deeper levels of errors. All I need to know as a developer is what went wrong and where it went wrong, and if possible the immediate cause of the problem.
2. Service actions that return their errors in their resulting output object rather than by throwing exceptions. The fact that it seems to be random which any given function will do is particularly frustrating. The SDK should smooth this over by throwing exceptionos for these instead.

**Comment by ganeshnj** [comment]
![image](https://user-images.githubusercontent.com/8882380/228063462-09a59ae9-0189-4324-bc36-89b2ffa634ec.png)

Experience for generic SDKs
```swift
let client = WeatherClient()
let input = GetWeatherInput(city: "seattle")
do {
    let output = try client.getWeather(input: input)    
} catch let error as WeatherError {
    print("Weather Service Error: \(error)")
} catch let error as ClientError {
    print("Client Error: \(error)")
} catch {
    print("Unknown Error: \(error)")
}
```

AWS Service experience
```swift
let client = AWSS3Client()
let input = GetObjectInput(bucket: "my-bucket", key: "my-key")
do {
    let output = try client.getObject(input: input)
} catch let error as NoSuchKeyError {
    print("Key does not exist: \(error)")
} catch let error as AWSS3Error {
    print("S3 Service Error: \(error)")
} catch let error as ClientError {
    print("Client Error: \(error)")
} catch {
    print("Unknown Error: \(error)")
}
```

**Comment by shepazon** [comment]
@ganeshnj That looks fabulous to me. Makes it easy for developers to provide strong error handling, recovering automatically from errors where it's possible to do so, while finding graceful paths to escape the situation when available, so that aborting operations outright only happens in extreme cases rather than most cases.

**Comment by jbelkins** [comment]
Shipped in AWS SDK for Swift 0.18.0
